homeassistant/components/plugwise/entity.py

```diff
@@ -1,13 +1,42 @@
 """Generic Plugwise Entity Class."""
 from __future__ import annotations
 
+from collections.abc import Callable, Coroutine
+from typing import Any, TypeVar
+
+from plugwise.exceptions import PlugwiseException
+
 from homeassistant.const import ATTR_NAME, ATTR_VIA_DEVICE, CONF_HOST
+from homeassistant.exceptions import HomeAssistantError
 from homeassistant.helpers.entity import DeviceInfo
 from homeassistant.helpers.update_coordinator import CoordinatorEntity
 
 from .const import DOMAIN
 from .coordinator import PlugwiseData, PlugwiseDataUpdateCoordinator
 
+T = TypeVar("T")
+
+
+def plugwise_exception_handler(
+    func: Callable[..., Coroutine[Any, Any, T]]
+) -> Callable[..., Coroutine[Any, Any, T]]:
+    """Decorate Plugwise calls to handle Plugwise exceptions.
+
+    A decorator that wraps the passed in function, catches Plugwise errors,
+    and requests an coordinator update to update status of the devices asap.
+    """
+
+    async def handler(self, *args: Any, **kwargs: Any) -> T:
```

Comment:
This is our normal way of typing these kind of decorators nowadays:
https://github.com/home-assistant/core/blob/a6013dc0de5c04e8425cf9ce3c7409f096de8fab/homeassistant/components/evil_genius_labs/util.py#L12-L29

Reply:
ParamSpec = Python 3.10...

Reply:
Oh jeez typing_extensions is back... ok.

Still unsure why that ignore would be needed, is that a mypy issue?

Reply:
I think it's because Concatenate is not supported yet by mypy:
http://mypy-lang.blogspot.com/2021/12/mypy-0930-released.html

@cdce8p knows best.

Reply:
Ahh its the concatenate. 

Not sure If I'm a fan of the ignore, but I guess that resolves itself in the future. Otherwise pretty cool to preserves the original signature.

Will update 👍

Reply:
Makes me wonder and now inspired by the evil genius one.

As this decorator is used to wrap command to Plugwise and handle exceptions, maybe I should rename it to `plugwise_command` and always schedule the coordinator update request.

And remove the "fake" state setting in the entity platforms when a change is done (by e.g., the turn_off command of a switch).

What do you think @MartinHjelmare?

Reply:
I think we should always request a coordinator update. If we should apply an assumed intermediary local state or not until the state is refreshed depends on the API. It's best if we don't need that, but sometimes we may do.

Reply:
Right thanks, in that case I'll start with moving the coordinator request refresh into the finally block to ensure it is called and rename the decorator.

Thanks for the input 👍

Reply:
> I think it's because Concatenate is not supported yet by mypy:

Yeah, we are still working on the mypy implementation. Some last details to hash out. The hope is to have it ready for the next release. The syntax with `ParamSpec` and double `Concatenate` to type `self` is correct though.

> Not sure If I'm a fan of the ignore, but I guess that resolves itself in the future. Otherwise pretty cool it preserves the original signature.

The tradeoff is worth it I think. The mypy output didn't change, and with the new typing we at least get better intellisense support in VS Code. It'll also be quite helpful when mypy starts to support it, just remove the ignores fix a few last errors and we're done.